### PR TITLE
Font Library: Font Collections frontend

### PR DESCRIPTION
**Pull request by matiasbenedetto**
## What?
Add Font Collections frontend

## Why?
To allow extenders to add new font collections

## How?
Adding frontend code to consume font collections from the backend

## Testing Instructions
1. Navigate to the font library modal
2. Go to Google Fonts Tab
3. Try to install Google Fonts

##  Test in Playground
[Test this PR in WordPress Playground](https://playground.wordpress.net/#{%22landingPage%22:%22/wp-admin/%22,%22steps%22:[{%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22},{%22step%22:%22mkdir%22,%22path%22:%22/wordpress/pr%22},{%22step%22:%22writeFile%22,%22path%22:%22/wordpress/pr/pr.zip%22,%22data%22:{%22resource%22:%22url%22,%22url%22:%22/plugin-proxy?org=WordPress&repo=gutenberg&workflow=Build%20Gutenberg%20Plugin%20Zip&artifact=gutenberg-plugin&pr=54566%22,%22caption%22:%22Downloading%20Gutenberg%20PR%2054566%22},%22progress%22:{%22weight%22:2,%22caption%22:%22Applying%20Gutenberg%20PR%2054566%22}},{%22step%22:%22unzip%22,%22zipPath%22:%22/wordpress/pr/pr.zip%22,%22extractToPath%22:%22/wordpress/pr%22},{%22step%22:%22installPlugin%22,%22pluginZipFile%22:{%22resource%22:%22vfs%22,%22path%22:%22/wordpress/pr/gutenberg.zip%22}}]}).

## Disclaimer
- `google-fonts.json` file ideally will be removed when https://github.com/WordPress/gutenberg/issues/53992 is merged.
- SVG previews are loaded [from this repo provisionally](https://github.com/matiasbenedetto/google-fonts-to-wordpress-collection/). This will be loaded from wordpress.org soon.

Implements: https://github.com/WordPress/gutenberg/issues/54040 https://github.com/WordPress/gutenberg/issues/54039 https://github.com/WordPress/gutenberg/issues/53882

### CODE REVIEW

**Comment by bph** [comment]
Tested via Playground these three steps from the description. 

- Navigate to the font library modal
- Go to Google Fonts Tab
- Try to install Google Fonts

It worked as described. What was missing was some kind of success message. The only indication that the font installed was that the **Install** button disappeared.

When I closed the Font management modal, I saw the fonts in the list of available fonts. 

Here is the short video of my test. 

https://github.com/WordPress/gutenberg/assets/39980/eefcae6d-e2c8-48aa-a176-6791f40de61a

**Comment by jasmussen** [comment]
Took it for a spin, very nice, here's a GIF showing me opening the manage fonts modal, clicking the Google Fonts tab, scrolling, picking a font, adding it, closing the modal, then saving:

google fonts process

A few things. Is the "confirm that you want to connect to Google's servers" notice built yet? I did not see it. This is likely the primary blocker to get solved before we merge.

We used to have a Upload tab. With a Google fonts tab, especially one that requires approval, it might be worth restoring that tab. It's not a blocker for this PR at all, but we do have a mockup for it:

upload mockup

You can still make it so dropping a font file onto the modal itself, regardless of tab, uploads. But it's still valuable to have the separate tab with the upload dropzone/button combo.

The "Google Fonts" tab, I think we discussed separately calling it "Install Fonts", didn't we? And we can improve the header area here a bit:

Screenshot 2023-09-19 at 10 09 15

Most of this polish can be followup stuff, post-beta 1, so also not a blocker, but here's the mockup to aim for:

header area

Also not a blocker: when a Google font is already installed, is it possible to show the box checked when returning to its permalink? Note here the Aclonica font both in the modal, and in the sidebar on the right:

Screenshot 2023-09-19 at 10 11 54

Here's a flow that could be improved a bit (also post beta-1). On this screen I've checked the box to install the font "Abril Fatface" from Google fonts:

google fonts

When I click the "Install" button in that screenshot above, I am taken to the following screen, which is exactly the same screen as before (Google Fonts → Abril Fatface), except now the box is just unchecked:

google fonts box unchecked

This is a bit confusing. We should do one of two things. First option: clicking the Install button can close the modal. Maybe it can even open the multi entity saving flow, showing you this screen:

multi entity saving

**Or**, clicking the Install button could take you to the _Library_ tab, with the newly added Google Font. This screen:

library permalink

Either of those, IMO, would be valid enough.

Unrelated to this PR (also not a blocker), but I think I mentioned we should consider not allowing you to disable the system font (or even any font stacks that are not _font files_). We can even hide them from the interface potentially. Shown here, the "System fonts" stack, in both the modal and the sidebar:

system fonts in sidebar
system fonts in modal

Really nice work, I think we can land this in short order. As noted, I think **the only blocker** is ensuring the Google server connection notice is in place. Renaming the tab from "Google Fonts" to "Install Fonts", is probably close to a blocker, but should also be easy.

File: lib/experimental/fonts/font-library/class-wp-font-collection.php
Comment (jffng) [inline lib/experimental/fonts/font-library/class-wp-font-collection.php]: I think this change is causing the `wpFontCollection/getData` test to fail.

```
--- Expected
+++ Actual

-    'data' => '{"this is mock data":true}'
+    'data' => stdClass Object &000000000227801400000000191e93fe (
+        'this is mock data' => true
+    )
```

File: lib/experimental/fonts/font-library/font-library.php
Comment (jffng) [inline lib/experimental/fonts/font-library/font-library.php:66]: Since there is now a default font collection, I think [this test](https://github.com/WordPress/gutenberg/blob/f3a6ff10f1c2055fdf4f62d2e851778d1a0f6db2/phpunit/tests/fonts/font-library/wpFontLibrary/getFontCollections.php#L38) needs to be updated — specifically the count of collections to expect after registering additional collections.

**Comment by jffng** [comment]
## Test Report

### Steps to Test

1. From site editor, visit the *Styles > Typography > Google Fonts* to ensure the Google fonts collection appears as expected.
2. Filter fonts using the Category dropdown 
3. Search fonts using the Search text input
4. Select a font, install and activate a single variant, ensure it appears in editor and view
5. Select a font, install and activate multiple variants, ensure they appear in editor and view
6. Delete a font family added via the Google Fonts collection

### Environment
- OS: macOS 13.0.1
- Browser: Google Chrome 116.0.5845.187
- PHP: 8.0.0
- WordPress: 6.3.1
- Theme: twentytwentyfour trunk
- Active Plugins:
	- gutenberg this branch

### Actual Results
**✅  1 - 3. Viewing, searching, filtering**

**✅  4. Installing a single variant**
Note: the font face does not appear in the editor right away, but requires a page refresh. I think this is a documented issue already?

**❌ 5. Installing multiple variants**

I discovered a small bug:

- Navigate to a font family that contains more than one variant and contains multiple font _styles_ (e.g. Advent Pro)
- Attempt to select a font face that has another variant with different style but same weight (e.g. Advent Pro 100 Italic). 
- Verify that selecting and de-selecting the target (via keyboard or mouse) changes the order it appears in the list:

https://github.com/WordPress/gutenberg/assets/5375500/12149645-37fb-4884-91d7-d0979d975801

The variants that are selected are in fact installed, so that aspect is working. 
 
**✅ 6. Delete a font family**

**Comment by matiasbenedetto** [comment]
> I think the only blocker is ensuring the Google server connection notice is in place

@jasmussen added:

Screenshot from 2023-09-19 14-17-41

**Comment by matiasbenedetto** [comment]
>  Renaming the tab from "Google Fonts" to "Install Fonts", is probably close to a blocker, but should also be easy.

@jasmussen I don't think this is the best alternative. What do you think of using the name of the font collections as the title of those tabs?

File: lib/experimental/fonts/font-library/font-library.php
Comment (matiasbenedetto) [reply lib/experimental/fonts/font-library/font-library.php:66]: That test tests the capabilities to register font collections and counts the 2 mock collections registered in the test. It is not testing the inclusion of the Google fonts collections.

**Comment by matiasbenedetto** [comment]
> ❌ 5. Installing multiple variants
> To reproduce, select a font family that contains more than one variant, and attempt to select a font face. Verify that the target selected (via keyboard or mouse) changes the list order of the variants

@jffng I'm not able to reproduce this. Could you add more details about how to reproduce this, please?

**Comment by jasmussen** [comment]
> I don't think this is the best alternative. What do you think of using the name of the font collections as the title of those tabs?

I don't recall where we discussed this, but I believe it was @mtias who specifically requested "Install Fonts". I wouldn't let it be blocking to this PR, but I think it's pretty important we follow a pretty straight beeline and not diverge too far from what gets decided. We can always change it so if there's an extensibility tab, we change the title at that point.

**Comment by jffng** [comment]
@matiasbenedetto yes, I added some details to the original report.

**Comment by matiasbenedetto** [comment]
> ❌ 5. Installing multiple variants
To reproduce, select a font family that contains more than one variant, and attempt to select a font face. Verify that the target selected (via keyboard or mouse) changes the list order of the variants

Thanks for the details @jffng . I couldn't reproduce this because I was testing on Firefox, and it was working as expected there.
It was a problem related a non-needed sort function. It should be fixed now :heavy_check_mark:

File: lib/experimental/fonts/font-library/font-library.php
Comment (ironprogrammer) [inline lib/experimental/fonts/font-library/font-library.php]:
```suggestion
	'description'    => __( 'Add from Google Fonts. Fonts are copied to and served from your site.' ),
```
Fixes typo and uses more user-friendly (?) "site" rather than "server".

File: packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js
Comment (ironprogrammer) [inline packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js]:
```suggestion
							'You can alternatively upload files directly on the Library tab.'
```
Possible refinement.

File: packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js
Comment (ironprogrammer) [inline packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js]:
```suggestion
							'To install fonts from Google you must give permission to connect directly to Google servers. The fonts you install will be downloaded from Google and stored on your site. Your site will then use these locally-hosted fonts.'
```
I'm sure this will go through multiple iterations, but here's a stab at refinement.

File: packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js
Comment (ironprogrammer) [inline packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js:42]: Will the fonts management dialog also have a way to _opt out_ of this access after it had been previously granted?

**Review by ironprogrammer — COMMENTED** [review]
This is looking/functioning awesome! Here are a few textual suggestions for consideration.

File: lib/experimental/fonts/font-library/font-library.php
Comment (gziolo) [inline lib/experimental/fonts/font-library/font-library.php]: > google-fonts.json file ideally will be removed when https://github.com/WordPress/gutenberg/issues/53992 is merged.

It would be great to avoid maintaining the JSON file in the repository that is over 50 kB.

File: packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js
Comment (gziolo) [inline packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js]: In other places, the local storage is accessed through the `window` global. This way, it won't need to be listed as a global in the ESLint config.

File: packages/eslint-plugin/configs/recommended-with-formatting.js
Comment (gziolo) [inline packages/eslint-plugin/configs/recommended-with-formatting.js]: Let's avoid exposing another global. In other places, it's done through `window.localStorage`.

**Comment by ironprogrammer** [comment]
## Test Report

### Environment
- OS: macOS 13.5.2
- Browser: Safari 16.6
- Server: nginx/1.25.2
- PHP: 7.4.33
- WordPress: 6.4-alpha-56267-src
- Theme: twentytwentythree v1.2
- Active Plugins:
  - gutenberg v16.6.0

### Actual Results
*(Test steps adapted from @jffng's [report above](https://github.com/WordPress/gutenberg/pull/54566#issuecomment-1726031821).)*

**1 - 3. Viewing, searching, filtering**
✅ Search and category filtering works as expected.

<details>
<summary>⚠️ Some font previews are clipped or not rendered legibly, so might be hard to decipher. Expand this section to see the list of items (this really gave my tired eyes a workout checking these 😅).</summary>

- Chenla
- Coiny
- Concert One
- Content
- Cutive
- Englebert
- Khmer
- Material Icons
- Material Icons Outlined
- Material Icons Round
- Material Icons Sharp
- Material Icons Two Tone
- Noto Color Emoji
- Noto Emoji
- Noto Kufi Arabic
- Noto Sans Arabic
- Noto Sans Bhaiksuki
- Noto Sans Caucasian Albanian
- Noto Sans Cypriot
- Noto Sans Elymaic
- Noto Sans Gunjala Gondi
- Noto Sans Lycian
- Noto Sans Myanmar
- Noto Sans Old Hungarian
- Noto Sans Old Turkic
- Noto Sans Phags Pa
- Noto Sans SignWriting
- Noto Sans Warang Citi
- Noto Serif Dogra
- Noto Serif Myanmar
- Revalia
- Siemreap

</details>

**4. Installing**
- ✅ Fonts can be installed and appear in the UI.
- ✅ During the Google font install flow, only the requested font file was downloaded from Google's servers.
- ❌ Installed Google font files are not being saved to `/wp-content/fonts/` as expected.
- ❌ After refreshing the page and observing activated font faces output to `#wp-fonts-local`, the Google font definition URLs reference the original `fonts.gstatic.com` locations. It is expected that each installed font would be downloaded and served locally. This occurs in both the editor and frontend of the site. Here's an example from `#wp-fonts-local`:
```css
@font-face {
    font-family: "Zen Dots";
    font-style: normal;
    font-weight: 400;
    font-display: fallback;
    src: url('https://fonts.gstatic.com/s/zendots/v12/XRXX3ICfm00IGoesQeaETM_FcCIG.ttf') format('truetype');
}
```

**5. Activating**
- ✅ On the Library tab, activated Google fonts display as checked.
- ⚠️ On the Google Fonts tab, after checking boxes for and installing Google fonts, the tab refreshes and the fonts appear *unchecked* (as described by @jasmussen above). The same happens when navigating back to an already-installed font through the Google Fonts tab.
- ⚠️ Fonts listed in the Typography sidebar only reflect active weights/styles when the corresponding variants are printed to `#wp-fonts-local` (i.e. after page refresh). Toggling activation of non-printed variants doesn't have an effect. *See Figure 1.*
- ✅ Checking multiple Google font variants does not re-sort items in list (originally [reported](https://github.com/WordPress/gutenberg/pull/54566#issuecomment-1726031821) by @jffng, now fixed).

**6. Deleting**
- ✅ Deleting Google fonts works as expected.
- ⚠️ After manually uploading a variable Oswald WOFF2 font, I installed the same font's six individual TTF variants using the Google Fonts tab. All were listed together under "Oswald". This could complicate things for users who wished to remove either the Google or manually uploaded fonts.

### Additional Notes
- Should non-alphabetic font sets include a friendly name, such as for Flow Block, Redacted, or Wavefont? For instance, compare with the [Flow Block search result](https://fonts.google.com/?query=flow+block) on Google Fonts.
- Scrolling all the way through the fonts appears to have downloaded about 55 MB of data to the page (mostly SVGs).
- The loading of font previews is fast, and the scrolling smooth 👍
- Items marked with ⚠️ above should be considered non-blocking for intents of this report, to be iterated on separately.

### Supplemental Artifacts
*Figure 1: Fonts displayed in Typography panel do not display accurately until after a page refresh.*
  Fonts listed in Typography panel do not display accurately until after a page refresh

File: packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js
Comment (matiasbenedetto) [reply packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js:31]: update wording

File: packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js
Comment (matiasbenedetto) [reply packages/edit-site/src/components/global-styles/font-library-modal/google-fonts-confirm-dialog.js:42]: currently nope

File: packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js
Comment (matiasbenedetto) [reply packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js]: done

**Comment by matiasbenedetto** [comment]
> ❌ Installed Google font files are not being saved to /wp-content/fonts/ as expected.
> ❌ After refreshing the page and observing activated font faces output to #wp-fonts-local, the Google font definition URLs reference the original fonts.gstatic.com locations. It is expected that each installed font would be downloaded and served locally. This occurs in both the editor and frontend of the site. Here's an example from #wp-fonts-local:

Fixed, it should be working as expected now.

**Comment by matiasbenedetto** [comment]
> ⚠️ After manually uploading a variable Oswald WOFF2 font, I installed the same font's six individual TTF variants using the Google Fonts tab. All were listed together under "Oswald". This could complicate things for users who wished to remove either the Google or manually uploaded fonts.

I think this is the expected behavior. As in your desktop computer when you install fonts, no matter the source of them all are grouped by font family and not by source of download.

File: lib/experimental/fonts/font-library/font-library.php
Comment (matiasbenedetto) [reply lib/experimental/fonts/font-library/font-library.php]: yes that will be addressed when that PR is merged.

File: lib/experimental/fonts/font-library/class-wp-font-collection.php
Comment (matiasbenedetto) [reply lib/experimental/fonts/font-library/class-wp-font-collection.php]: Thanks for flagging. It should be fixed now.

File: lib/experimental/fonts/font-library/font-library.php
Comment (matiasbenedetto) [reply lib/experimental/fonts/font-library/font-library.php:66]: You were right as we are using a static class there is only one list of font collections. I refactored the tests to work as expected.

**Comment by matiasbenedetto** [comment]
> Should non-alphabetic font sets include a friendly name, such as for Flow Block, Redacted, or Wavefont? For instance, compare with the [Flow Block search result](https://fonts.google.com/?query=flow+block) on Google Fonts.

Yes, that was accidentally lost. Added again:

image

File: packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js
Comment (gziolo) [reply packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js]: thank you!

**Comment by jasmussen** [comment]
Nice work. Took this for a spin, and I think we should merge this as is. 

There are some [followups here that I would recommend as next steps](https://github.com/WordPress/gutenberg/pull/54566#issuecomment-1725042244), notably deciding what happens when you uncheck a box and press "Update". But also as noted, that should not block this PR. 

Also some things to follow up on:

Screenshot 2023-09-20 at 17 24 52

The line-height is a bit tight inside this box. I don't know if the Text component allows tweaking that without adding custom CSS here, not sure, perhaps @ciampo knows? But it's a bit tight.

The padding should also be a uniform 32px all around, and the blue button centered, as in the mockup:

Screenshot 2023-09-20 at 17 25 40

And finally, can we remove the "Default font collection" subheading here?

Screenshot 2023-09-20 at 17 24 58

**Review by jasmussen — APPROVED** [review]
I recommend merging this as is, and keeping all the polish separate, that can certainly wait. Nice work!

**Comment by ironprogrammer** [comment]
Thanks for the quick turn on these updates, @matiasbenedetto! I'd like to echo the [input provided by @jasmussen above](https://github.com/WordPress/gutenberg/pull/54566#issuecomment-1727957500) regarding post-merge follow-ups for the the access modal and subheading. This looks good to go! 🚀 

## Test Report

### Actual Results
*(This report is a follow-up to critical items in my [previous report](https://github.com/WordPress/gutenberg/pull/54566#issuecomment-1727127645).)*

**4. Installing**
- ✅ ✅ During the Google font install flow, I did not see any requests to `fonts.gstatic.com` (privacy FTW!), implying that client browser data was not shared with the 3rd-party domain. Installed Google font files are now requested from local URLs (**improved since previous report**).
- ✅ Installed Google font files are saved to `/wp-content/fonts/` (**fixed since previous report**).
- ✅ After refreshing the page, installed Google fonts appear in `#wp-fonts-local` using local URL `src: url()`s. This was confirmed in both the editor and frontend of the site. Here's an example (**fixed since previous report**):
```css
@font-face {
    font-family: Redacted;
    font-style: normal;
    font-weight: 400;
    font-display: fallback;
    src: url('http://wp-src.test/wp-content/fonts/redacted_normal_400.ttf') format('truetype');
}
```

### Additional Notes
- Noting that the local storage value `wp-font-library-google-fonts-permission` was swapped for `wp-font-library-default-font-collection-permission`, so I was re-prompted to allow access to 3rd-party servers 👍.

**Comment by bph** [comment]
It would be great to have this in 6.4 beta 1 and in Gutenberg 16.7 -

**Comment by ciampo** [comment]
> The line-height is a bit tight inside this box. I don't know if the Text component allows tweaking that without adding custom CSS here, not sure, perhaps @ciampo knows? But it's a bit tight.

I'm not very familiar with the `Text` component, but scanning the code it seems like:

- it accepts a `lineHeight` prop (same values as in CSS)
- otherwise, it has a `adjustLineHeightForInnerControls` which tries to match the line height when the text is rendered side-to-side with a control (side note: the values there seem to be out of date and would probably benefit from a refresh using the latest values)

In short, I'm not sure about the `Text` component being particularly helpful in this case.

**Comment by mikachan** [comment]
I just cherry-picked this PR to the release/16.7 branch to get it included in the next release: b4a28f01ea

**Pull request merged**
